fasthttp_test.go

```diff
@@ -433,6 +435,30 @@ func TestFastHTTPNodeApplyMiddlewareStatic(t *testing.T) {
 	}
 }
 
+func TestFastHTTPNodeApplyMiddlewareDeepStatic(t *testing.T) {
+	t.Parallel()
+
+	router := NewFastHTTPRouter().(*fastHTTPRouter)
+
+	router.GET("/x/x/{param}", func(ctx *fasthttp.RequestCtx) {
+		if _, err := fmt.Fprintf(ctx, "x"); err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	router.USE(http.MethodGet, "/x/x", mockFastHTTPMiddleware("m1"))
+	router.USE(http.MethodGet, "/x/x/{param}", mockFastHTTPMiddleware("m2"))
+	router.USE(http.MethodGet, "/x/x/x", mockFastHTTPMiddleware("m3"))
+
+	ctx := buildFastHTTPRequestContext(http.MethodGet, "/x/x/x")
```

Comment:
Test this case, to have an idea of an order I am talking about.
Without any changes you did I suspect the order to be `m1->m2->mx1->mx2->mxy1->mxy2->mparam1->mparam2->handler` Where the static handler group for a wildcard route (`mxy1->mxy2->`) comes before wildcard middleware for the same route (`mparam1->mparam2->`) and they were applied in different order (wildcard group before static). Other middleware order is intact (should be).

And with your changes every middleware will be reordered. Possibly giving output like this: `mparam2->mparam1->mxy2->mxy1->mx2->mx1->m2->m1->handler`.

```suggestion
 	router.GET("/x/{param}", func(ctx *fasthttp.RequestCtx) {
 		if _, err := fmt.Fprintf(ctx, "handler"); err != nil {
 			t.Fatal(err)
 		}
 	})

 	router.USE(http.MethodGet, "/", mockFastHTTPMiddleware("m1->"))
 	router.USE(http.MethodGet, "/", mockFastHTTPMiddleware("m2->"))

 	router.USE(http.MethodGet, "/x", mockFastHTTPMiddleware("mx1->"))
 	router.USE(http.MethodGet, "/x", mockFastHTTPMiddleware("mx2->"))

 	router.USE(http.MethodGet, "/x/{param}", mockFastHTTPMiddleware("mparam1->"))
 	router.USE(http.MethodGet, "/x/{param}", mockFastHTTPMiddleware("mparam2->"))

 	router.USE(http.MethodGet, "/x/y", mockFastHTTPMiddleware("mxy1->"))
 	router.USE(http.MethodGet, "/x/y", mockFastHTTPMiddleware("mxy2->"))

 	ctx := buildFastHTTPRequestContext(http.MethodGet, "/x/y")

 	router.HandleFastHTTP(ctx)

  	if string(ctx.Response.Body()) != "m1->m2->mx1->mx2->mxy1->mxy2->mparam1->mparam2->handler" {
 		t.Errorf("Use middleware error %s", string(ctx.Response.Body()))
 	}
```

Reply:
Thanks for this clarification and example test. I think now I understand better what you are describing. It's the default sorting order applied in withNode that defines which comes first and which after. That's why reverse can not solve it. 

How about if we keep track of middleware order index by the way router.USE lines were ordered ? 
Then we create a function normalizing the tree by the middleware order index ? 

I understand that isMoreIMportant ordering is critical to the build of the tree but we can place this reordering by middleware index say:

```
if len(orphanMatches) > 0 {
        for i := 0; i < len(orphanMatches); i++ {
		m = orphanMatches[i].node.Middleware().Merge(m)
	}
        // Sort Nodes by middleware order index
	sort.Slice(t, func(i, j int) bool {
		return isMiddlewareOrder(t[i], t[j])
	})
}
```

That has a few advantages:
* the final order returned is not limited to the default gorouter order but defined by the order of the defined middleware
* it is out of the for loop so done only once
* we can keep an index in router.USE and use it for isMiddlewareOrder

Reply:
It is not as simple as that. Because even if we track the order of middleware in which it was applied here. There is going to be the issue that the middleware for node **with a route** has already merged middleware of a parent nodes.

Reply:
what we could do is move the orphan middleware to be merged at the and of routes middleware. in most cases its going to be some static path orphan middleware that needs to be applied to a wildcard route like in our example. 

and then i think we could leave it like that

Reply:
I see, yes we can try that. Basically for anything one or two levels deep my basic needs are to have predictable ordering for all types of nodes. If we have to guess or document it as a quirk that's kind of becoming like the weak spot of the go-api-boilerplate where the most limitations apply.